Document internal Audio Architecture (#7896)
godotengine/godot-docs: contributing/development/core_and_modules/internal_audio_architecture.rst

.. _doc_internal_audio_architecture:

Internal audio architecture
===========================

This page is a high-level overview of Godot 4's internal audio design.
It does not apply to previous Godot versions.

The goal of this page is to document design decisions taken to best suit
:ref:`Godot's design philosophy <doc_best_practices_for_engine_contributors>`,
while providing a starting point for new audio contributors.

If you have questions about audio internals not answered here, feel free to
ask in the ``#audio`` channel of the
`Godot Contributors Chat <https://chat.godotengine.org/channel/audio>`__.

> 
> 
> .. note::
> 
>     If you have difficulty understanding concepts on this page,
>     it is recommended to go through an audio blog such as
>     `Audio Sample Rate and Bit Depth <https://www.izotope.com/en/learn/digital-audio-basics-sample-rate-and-bit-depth.html>`__,
>     `How Sound Works <https://audiouniversityonline.com/audio-basics-how-sound-works/>`__, and
>     `An Introduction to Audio Electronics <https://www.allaboutcircuits.com/technical-articles/introduction-audio-electronics-sound-microphones-speakers-amplifiers>`__.
> 
> 

Reviewer: Use 1 blank line before and after (instead of 2):

```suggestion

.. note::

    If you have difficulty understanding concepts on this page,
    it is recommended to go through an audio blog such as
    `Audio Sample Rate and Bit Depth <https://www.izotope.com/en/learn/digital-audio-basics-sample-rate-and-bit-depth.html>`__,
    `How Sound Works <https://audiouniversityonline.com/audio-basics-how-sound-works/>`__, and
    `An Introduction to Audio Electronics <https://www.allaboutcircuits.com/technical-articles/introduction-audio-electronics-sound-microphones-speakers-amplifiers>`__.

```

.. _doc_internal_audio_architecture_drivers:

> Audio Drivers
> -------------

Reviewer: Use sentence case for headings:

```suggestion
Audio drivers
-------------
```

Supported

> - Dummy

Reviewer: ```suggestion
- Dummy (no audio output, used for headless mode)
```

> - linuxBSD

Reviewer: ```suggestion
- Linux/\*BSD
```

>    - Alsa

Reviewer: ```suggestion
   - ALSA
```

   - PulseAudio
   - Alsa MIDI

- Windows + UWP

   - WASAPI

>    - XAudio2

Reviewer: XAudio2 isn't compiled by default, and [it currently doesn't build](https://github.com/godotengine/godot/issues/75109#issuecomment-1724273758). It may be removed from the source tree at some point.

   - Win MIDI

- macOS + iOS

   - CoreAudio
   - Core MIDI

- Android

   - OpenSL ES

.. _doc_internal_audio_architecture_codecs:

Audio Codecs
------------

OGG Vorbis
^^^^^^^^^^

.. note::

>     This is an audio format that takes a bit longer than WAV to play,
>     but is smaller in file size than WAV and MP3. Therefore most suitable for playing music in-game.

Reviewer: ```suggestion
    This is an audio format that takes more processing power than WAV to play back,
    but is smaller in file size than WAV and MP3. It's most suitable for
    playing background music.
```

**Why not vorbisfile?**
// TODO

WAV
^^^

.. note::

>     This is an audio format that is the qickest to load, but is larger in file size than Ogg Vorbis and MP3. Therefore most suitable for playing sound effects.

Reviewer: ```suggestion
    This is an audio format that requires the least processing power to play back,
    but is larger in file size than Ogg Vorbis and MP3. It's most suitable for
    playing short sound effects.
```

MP3
^^^

.. _doc_internal_audio_architecture_servers:

> AudioServer
> ------------

Reviewer: Title underline must match the title's length (use a fixed-width font in your editor to notice this more easily):

```suggestion
AudioServer
-----------
```

Author: Thanks for catching that.  
I do use a monospaced font, but I copy&paste the underlines because the underline typing is annoying.

   - side-right speaker

But those are not the total channels, as each has a left and right side (i.e. left and right speaker).
The total channels are twice the size mentioned above and equals the number of speakers.
Depending on the context, one or the other size logic is used.

The speaker mode is sometimes driver-dependent. All drivers lookup the mode by using the total number of channels.
Except Alsa, XAudio2 and OpenSL ES. These enforce stereo speaker mode and support only 2 total channels.

**Bus**
Godot as an internal bus logic to separately control volume and audio effects applied to audio streams.
All busses eventually feed into the master bus, which is output to the audio output device.

Seeking / Looping
^^^^^^^^^^^^^^^^^

SpatialSoundServer
^^^^^^^^^^^^^^^^^^

SpatialSound2DServer
^^^^^^^^^^^^^^^^^^^^

.. _doc_internal_audio_architecture_classes:

Core audio classes architecture
-------------------------------

.. _doc_internal_audio_architecture_2d_vs_3d:

2D and 3D audio separation
--------------------------

.. _doc_internal_audio_architecture_techniques:

2D audio techniques
-------------------

3D audio techniques
-------------------

> 

Reviewer: It's a good idea to mention SPCAP, Camera3D doppler tracking and the audio panning properties here.

Author: Copy&pasted that as-is into the file for now. I'll write that part today.
